File: frontend/src/engine/capabilities/addDropdownControl.ts
/**
 * addDropdownControl – Create a dropdown (data validation list) in a cell.
 *
 * The dropdown can be driven by a comma-separated list of values or by a
 * range reference (e.g. "Sheet2!A:A"). Useful as a filter control or
 * user-input mechanism.
 *
 * Office.js notes:
 * - DataValidation API is in ExcelApi 1.8+.
 * - For list validation, formula1 is either a comma-separated string or a
 *   range reference prefixed with "=".
 */

import { CapabilityMeta, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange, stripWorkbookQualifier } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "addDropdownControl",
  description: "Create a dropdown (data validation list) in a cell",
  mutates: true,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.8",
};

/**
 * Determine whether a listSource string looks like a range reference
 * rather than a comma-separated list of literal values.
 */
function isRangeReference(source: string): boolean {
  const trimmed = source.trim();
  // Contains "!" (sheet-qualified) or ":" (multi-cell range) or is a simple column/cell ref
  if (trimmed.includes("!") || trimmed.includes(":")) return true;
  // Matches a simple cell or column reference like "A1", "B", "AA10"
  if (/^[A-Z]{1,3}\d*$/i.test(trimmed)) return true;
  return false;
}

async function handler(
  context: Excel.RequestContext,
  params: any,
  options: ExecutionOptions
): Promise<StepResult> {
  const {
    cell,
    listSource,
    promptMessage,
    sheetName,
  } = params;

  if (!cell || !listSource) {
    return {
      stepId: "",
      status: "error",
      message: "cell and listSource are required",
    };
  }

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would add dropdown control to ${cell}`,
    };
  }

  options.onProgress?.(`Adding dropdown control to ${cell}...`);

  // Resolve target cell — prepend sheetName if provided
  const cellAddress = sheetName && !cell.includes("!") ? `${sheetName}!${cell}` : cell;
  const range = resolveRange(context, cellAddress);

  // Build the validation source
  let source: string;
  let optionCount: number | string;

  if (isRangeReference(listSource)) {
    // Range reference — strip workbook qualifier and prefix with "="
    source = "=" + stripWorkbookQualifier(listSource);
    optionCount = "range " + listSource;
  } else {
    // Comma-separated literal values
    source = listSource;
    optionCount = listSource.split(",").length + " options";
  }

  const validationRule: Excel.DataValidationRule = {
    list: {
      inCellDropDown: true,
      source,
    },
  };

  range.dataValidation.rule = validationRule;

  if (promptMessage) {
    range.dataValidation.prompt = {
      showPrompt: true,
      title: "Select a value",
      message: promptMessage,
    };
  }

  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Added dropdown control to ${cell} with ${optionCount}`,
    outputs: { range: cell },
  };
}


// ── Legacy-Excel fallback (ExcelApi < 1.8) ────────────────────────────────────
// Range.dataValidation requires 1.8. We can't produce an actual in-cell
// dropdown arrow. The best we can do: write the allowed values to a helper
// section of the sheet and annotate the target cell so the user sees what
// values are expected. Status=success with a clear warning — data-entry
// correctness is the user's responsibility on old Excel versions.
async function fallback(
  context: Excel.RequestContext,
  params: any,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { cell, listSource, sheetName } = params;

  if (!cell || !listSource) {
    return { stepId: "", status: "error", message: "cell and listSource are required" };
  }

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would emit dropdown guidance for ${cell} (legacy fallback; no in-cell arrow).`,
    };
  }

  options.onProgress?.(`Legacy-Excel mode: data validation unavailable, annotating ${cell} instead...`);

  const cellAddress = sheetName && !cell.includes("!") ? `${sheetName}!${cell}` : cell;
  const anchor = resolveRange(context, cellAddress);

  // If the listSource is a literal comma-separated set, embed it in an inline
  // note next to the cell so the user sees the options. Range references are
  // kept as-is in the message.
  const rangeLike = isRangeReference(listSource);
  const optionsNote = rangeLike
    ? `Allowed values come from ${listSource}.`
    : `Allowed values: ${listSource}.`;

  // Write an inline annotation to the right-hand cell if empty.
  const noteCell = anchor.getOffsetRange(0, 1);
  noteCell.load(["values"]);
  await context.sync();

  const existing = noteCell.values?.[0]?.[0];
  if (existing === null || existing === "" || existing === undefined) {
    noteCell.values = [[`Dropdown: ${optionsNote}`]];
    noteCell.format.font.italic = true;
    noteCell.format.font.color = "#606060";
    noteCell.format.fill.color = "#F7F9FC";
    noteCell.format.wrapText = true;
    await context.sync();
  }

  return {
    stepId: "",
    status: "success",
    message:
      `Dropdown control at ${cell} could not be installed — data validation requires ExcelApi 1.8+. ` +
      `Inline annotation written instead. ${optionsNote} (legacy-Excel fallback).`,
    outputs: { range: cell },
  };
}

registry.register(meta, handler as any, fallback as any);
export { meta };

File: backend/app/execution/capabilities/add_dropdown_control.py
"""addDropdownControl — create a dropdown (list validation) on a cell using literal values or a range reference."""

from __future__ import annotations

import re
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


_V_LIST = 3       # xlValidateList
_ALERT_STOP = 1   # xlValidAlertStop

_SIMPLE_CELL_RE = re.compile(r"^[A-Z]{1,3}\d*$", re.IGNORECASE)
# Strip a leading [Book.xlsx] workbook qualifier from a range reference.
_WB_PREFIX_RE = re.compile(r"^\[[^\]]+\]")


def _is_range_reference(source: str) -> bool:
    trimmed = source.strip()
    if "!" in trimmed or ":" in trimmed:
        return True
    if _SIMPLE_CELL_RE.match(trimmed):
        return True
    return False


def _strip_workbook_qualifier(ref: str) -> str:
    """Remove `[Book.xlsx]` prefix from a range reference (Excel's Validation
    source can't reference an external workbook name in the formula)."""
    return _WB_PREFIX_RE.sub("", ref.strip())


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    cell = params.get("cell")
    list_source = params.get("listSource")
    prompt_message = params.get("promptMessage")
    sheet_name = params.get("sheetName")

    if not cell or not list_source:
        return {"status": "error", "message": "addDropdownControl requires 'cell' and 'listSource'."}

    # Qualify the cell address with sheetName when provided and not already qualified.
    cell_address = cell
    if sheet_name and "!" not in cell:
        cell_address = f"{sheet_name}!{cell}"

    rng = resolve_range(ctx.workbook_handle, cell_address)

    if ctx.dry_run:
        return {"status": "preview", "message": f"Would add dropdown control to {rng.address}."}

    # Build source and an informational count label.
    if _is_range_reference(list_source):
        source = "=" + _strip_workbook_qualifier(list_source)
        option_count_label = f"range {list_source}"
    else:
        source = list_source
        option_count_label = f"{len(list_source.split(','))} options"

    try:
        v = rng.api.Validation
        try:
            v.Delete()
        except Exception:
            pass
        v.Add(Type=_V_LIST, AlertStyle=_ALERT_STOP, Formula1=source)
        v.InCellDropdown = True
        if prompt_message:
            v.ShowInput = True
            v.InputTitle = "Select a value"
            v.InputMessage = prompt_message
    except Exception as exc:  # noqa: BLE001
        return {
            "status": "error",
            "message": f"Add dropdown failed: {exc}",
            "error": str(exc),
        }

    return {
        "status": "success",
        "message": f"Added dropdown control to {rng.address} with {option_count_label}.",
        "outputs": {"range": rng.address},
    }


registry.register("addDropdownControl", handler, mutates=True, affects_formatting=False)
